lib/plugins/aws/lib/naming.js

```diff
@@ -0,0 +1,231 @@
+'use strict';
+
+const _ = require('lodash');
+
+/**
+ * A centralized naming class that provides naming standards for the AWS provider plugin.  The
+ * intent is for this to enable naming standards to be configured using a custom naming plugin or
+ * a naming configuration feature, based on which of those is determined to in the best approach.
+ *
+ * There are two main concepts:
+ * 1. normalization - normalization of a name comprises removing any unacceptable characters and
+ * changing the name to comply with capitalization expectations.
+ * 2. logicalization - logicalization of a name comprises preparing it for use as an attribute name
+ * for a resource within the Service's CloudFormation Template.  Notably, a suffix or prefix should
+ * be added to the name of resources.  These help to disambiguate attributes identifying or
+ * declaring different aspects or data about the same resource.  (e.g. MyFuncLambdaFunction vs.
+ * MyFuncLambdaFunctionArn)  Generally, normalization of the name also occurs.
+ *
+ * Sometimes, it is important to deconstruct the names that are created via this centralized
+ * utility.  As a result, the regular expressions used for identifying specific resource types using
+ * their names or the logic to extract a name components from a logical name are things that must
+ * also be centralized so that they co-occur and are more correlated spatially within the code base.
+ * That is to say that they are easier to change together.
+ */
+class Naming {
+  constructor(serverless) {
+    this.serverless = serverless;
+  }
+
+  /*
+   * General Name Normalization
+   */
+  normalizeName(name) {
+    return `${_.upperFirst(name)}`;
+  }
+  normalizeNameToAlphaNumericOnly(name) {
+    return this.normalizeName(name.replace(/[^0-9A-Za-z]/g, ''));
+  }
+  normalizePathtoCapitalAlphaNumbericOnlyWithReplacement(path) {
+    return this.normalizeNameToAlphaNumericOnly(
+      path.replace(/-/g, 'Dash')
+          .replace(/\{(.*)\}/g, '$1Var'));
+  }
+
+  // Probably belongs in a non-aws-specific code asset
+  getServiceEndpointRegex() {
+    return /^ServiceEndpoint/;
+  }
+
+  /*
+   * Stack Naming
+   */
+  getStackName() {
+    return `${this.serverless.service.service}-${this.serverless.config.stage}`;
```

Comment:
@erikerikson about the stage/region not being available in the context, maybe in that case we could just pass the stage/region to whatever method needs them? this would make those methods "pure" in a sense. That would be similar to how we're doing it with `sdk.request` for example, you just pass it a stage/region. Do you think this setup would fix this current issue?

as for multiple providers, imo multi provider support in itself is not challenging...what's really challenging is **designing a uniform interface for all providers**, simply because each provider is completely different. I honestly am not completely sure how we're going to accomplish that since lots of other providers are still immature in their interface. But we're keeping our eyes open for any opportunities there 😊